test-data/unit/check-narrowing.test

```diff
@@ -1910,3 +1910,38 @@ if len(x) == a:
 else:
     reveal_type(x)  # N: Revealed type is "Union[Tuple[builtins.int, builtins.int], Tuple[builtins.int, builtins.int, builtins.int]]"
 [builtins fixtures/len.pyi]
+
+
+[case testNarrowingIsSubclassNoneType1]
+from typing import Type, Union
+
+def f(cls: Type[Union[None, int]]) -> None:
+    if issubclass(cls, int):
+        reveal_type(cls)  # N: Revealed type is "Type[builtins.int]"
+    else:
+        reveal_type(cls)  # N: Revealed type is "Type[None]"
+[builtins fixtures/isinstance.pyi]
+
+
+[case testNarrowingIsSubclassNoneType2]
+from typing import Type, Union
+
+def f(cls: Type[Union[None, int]]) -> None:
+    if issubclass(cls, type(None)):
+        reveal_type(cls)  # N: Revealed type is "Type[None]"
+    else:
+        reveal_type(cls)  # N: Revealed type is "Type[builtins.int]"
+[builtins fixtures/isinstance.pyi]
+
+
+[case testNarrowingIsSubclassNoneType3]
+from typing import Type, Union
+
+NoneType_ = type(None)
+
+def f(cls: Type[Union[None, int]]) -> None:
+    if issubclass(cls, NoneType_):
```

Comment:
One case is still broken, I believe: (python3.10+)

`if issubclass(cls, types.Nonetype)`

Reply:
You are right.  The definition in types.pyi is not helpful.  I assumed something similar to mentioned test:

```python
NoneType = type(None)
```
But it is:

```python
    @final
    class NoneType:
        def __bool__(self) -> Literal[False]: ...
```

My suggestion is to let `ExpressionChecker.analyze_ref_expr` directly return `TypeType(NoneType())` when it encounters the full name "types.NoneType".  (already pushed)